src/bentoml/_internal/marshal/dispatcher.py

```diff
@@ -172,6 +172,174 @@ async def controller(self):
         """
         A standalone coroutine to wait/dispatch calling.
         """
+        logger.debug("Starting dispatcher optimizer training...")
+        # warm up the model
+        while self.optimizer.outbound_counter <= self.optimizer.N_SKIPPED_SAMPLE:
+            try:
+                async with self._wake_event:  # block until there's any request in queue
+                    await self._wake_event.wait_for(self._queue.__len__)
+
+                n = len(self._queue)
+                now = time.time()
+                w0 = now - self._queue[0][0]
+
+                # only cancel requests if there are more than enough for training
+                if (
+                    n
+                    > self.optimizer.N_SKIPPED_SAMPLE
+                    - self.optimizer.outbound_counter
+                    + 6
+                    and w0 >= self.max_latency_in_ms
+                ):
+                    # we're being very conservative and only canceling requests if they have already timed out
+                    self._queue.popleft()[2].cancel()
+                    continue
+                # don't try to be smart here, just serve the first few requests
+                if self._sema.is_locked():
+                    await asyncio.sleep(self.tick_interval)
+                    continue
+
+                n_call_out = 1
+                # call
+                self._sema.acquire()
+                inputs_info = tuple(self._queue.pop() for _ in range(n_call_out))
+                self._loop.create_task(self.outbound_call(inputs_info))
```

Comment:
@sauyon  `create_task` doesn't mean `outbound_counter` will increase instantly. More requests will come into the loop. Do you expect this to happen?

Reply:
I'm basically waiting for the requests to finish before exiting this loop instead of having separate loops to wait for the requests to finish before triggering an optimizer refresh.

The main issue I'm working around here is the case that the runner errors; I'm not actually sure that was a real issue but I'm not sure how the training behaves in that case.

Reply:
I should probably check it again before we try to call out, though.